Make some methods of `Pin` unstable const

CDirkx opened the pull request:
Make the following methods unstable const under the `const_pin` feature:
- `new`
- `new_unchecked`
- `into_inner`
- `into_inner_unchecked`
- `get_ref`
- `into_ref`
- `get_mut`
- `get_unchecked_mut`

Of these, `into_inner` and `into_inner_unchecked` require the unstable `const_precise_live_drops`.

Also adds tests for these methods in a const context.

Tracking issue: #76654

r? @ecstatic-morse

CDirkx committed:
Make some methods of `Pin` unstable const

Make the following methods unstable const under the `const_pin` feature:
- `new`
- `new_unchecked`
- `into_inner`
- `into_inner_unchecked`
- `get_ref`
- `into_ref`

Also adds tests for these methods in a const context.

Tracking issue: #76654

CDirkx commented:
I remember checking if `Pin` methods could be made const before, however at that time I got an error that const fns could only be inside impl blocks that have `Sized` or no bounds on generic parameters. I couldn't find what enabled this?

CDirkx commented:
I excluded `into_ref` (implemented on on `Pin<&mut T>`), which could technically also be made const, as `&mut` in consts is still being figured out I believe.

ecstatic-morse commented:
> I remember checking if Pin methods could be made const before, however at that time I got an error that const fns could only be inside impl blocks that have Sized or no bounds on generic parameters. I couldn't find what enabled this?

This is currently part of the "min `const fn`" checks, which are disabled for const-unstable functions. Presumably you used `rustc_const_unstable` the last time you tried this.

> I excluded `into_ref` (implemented on on `Pin<&mut T>`), which could technically also be made const, as `&mut` in consts is still being figured out I believe.

I don't really see a reason to exclude this if it compiles. As long as these functions are all const-unstable, there's not really any risk.

cc @rust-lang/wg-const-eval Unlike `Option` and `Result`, I doubt `Pin` is useful in a const-context. However, making these unstably const might allow people to test things like const-qualification more thoroughly, and I don't see any downsides.

CDirkx commented:
I added the methods `into_ref`, `get_mut`, `get_unchecked_mut` from `Pin<&mut T>`, the tests for these are a bit different.

RalfJung commented on library/core/tests/pin.rs:
```diff
@@ -0,0 +1,32 @@
+    assert_eq!(REF, POINTER);
+
+    const fn pin_mut_const() {
+        // test that the methods of `Pin<&mut T>` are usable in a const context
+        // should be written as the other tests once `&mut` is usable in constants
```
Are these two sentences? I think they are. If yes, please add capitalization and `.`. I had to read this 5 times to successfully parse it.

ecstatic-morse commented:
r=me with Ralf's change.

@bors delegate+

bors commented:
:v: @CDirkx can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"ecstatic-morse","delegate":"CDirkx"} -->

CDirkx committed:
Make some methods of `Pin<&mut T>` unstable const

Make the following methods unstable const under the `const_pin` feature:
- `into_ref`
- `get_mut`
- `get_unchecked_mut`

CDirkx commented:
r=@ecstatic-morse

RalfJung commented:
Almost, the right command is

@bors r=ecstatic-morse

bors commented:
:pushpin: Commit e3c6e46168758642f0bab64da374f93ed21b1cd0 has been approved by `ecstatic-morse`

<!-- @bors r=ecstatic-morse e3c6e46168758642f0bab64da374f93ed21b1cd0 -->
<!-- homu: {"type":"Approved","sha":"e3c6e46168758642f0bab64da374f93ed21b1cd0","approver":"ecstatic-morse"} -->
